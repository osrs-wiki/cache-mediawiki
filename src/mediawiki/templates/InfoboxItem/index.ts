export { default as InfoboxItem } from "./InfoboxItem";
