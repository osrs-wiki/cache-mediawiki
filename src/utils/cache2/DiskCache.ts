import {
  ArchiveData,
  CacheProvider,
  CacheVersion,
  FileProvider,
  hash,
  IndexData,
} from "./Cache";
import { Reader } from "./Reader";

export class DiskIndexData implements IndexData {
  public id!: number;
  public protocol!: number;
  public revision!: number;
  public compression!: number;
  public crc!: number;
  public named!: boolean;
  /** @internal */ archives: Map<number, ArchiveData> = new Map();
}

export class DiskCacheProvider implements CacheProvider {
  private data: Promise<Uint8Array | undefined>;
  private indexData: Map<number, Promise<DiskIndexData | undefined>> =
    new Map();
  private pointers: Map<number, Promise<Reader | undefined>> = new Map();

  public constructor(private readonly disk: FileProvider) {
    this.data = disk.getFile("main_file_cache.dat2");
    this.getPointers(255);
  }

  public async getIndex(index: number): Promise<DiskIndexData | undefined> {
    let id = this.indexData.get(index);
    if (!id) {
      this.indexData.set(
        index,
        (id = (async () => {
          const cdata = await this.getArchiveRaw(255, index);
          if (!cdata) {
            return undefined;
          }
          const ad = new ArchiveData(255, index);
          ad.addFile(0, 0);
          ad.compressedData = cdata;
          const r = new Reader(ad.getFile(0)!.data);

          const out = new DiskIndexData();
          out.id = index;
          const protocol = (out.protocol = r.u8());
          out.revision = protocol >= 6 ? r.i32() : -1;

          const flags = r.u8();
          const named = (out.named = !!(flags & 1));

          const numArchives = protocol <= 6 ? r.u16() : r.u32o16();
          const ams: ArchiveData[] = new Array(numArchives);
          for (let i = 0, id = 0; i < numArchives; i++) {
            id += protocol <= 6 ? r.u16() : r.u32o16();
            const v = new ArchiveData(index, id);
            ams[i] = v;
            out.archives.set(id, v);
          }

          if (named) {
            for (const am of ams) {
              am.namehash = r.i32();
            }
          }
          for (const am of ams) {
            am.crc = r.i32();
          }
          for (const am of ams) {
            am.revision = r.i32();
          }

          const numFileses = new Uint32Array(ams.length);
          for (let i = 0; i < numArchives; i++) {
            numFileses[i] = protocol <= 6 ? r.u16() : r.u32o16();
          }

          for (let i = 0; i < numArchives; i++) {
            const am = ams[i];
            const numFiles = numFileses[i];
            let id = 0;
            for (let i = 0; i < numFiles; i++) {
              id += protocol <= 6 ? r.u16() : r.u32o16();
              am.addFile(id, 0);
            }
          }

          if (named) {
            for (const am of ams) {
              for (const file of am.files.values()) {
                (<any>file).namehash = r.i32();
              }
            }
          }

          return out;
        })())
      );
    }
    return id;
  }

  private async getPointers(index: number): Promise<Reader | undefined> {
    let ptrs = this.pointers.get(index);
    if (!ptrs) {
      this.pointers.set(
        index,
        (ptrs = (async () => {
          const data = await this.disk.getFile("main_file_cache.idx" + index);
          return data && new Reader(data);
        })())
      );
    }
    return ptrs;
  }

  public async getArchives(index: number): Promise<number[] | undefined> {
    const idx = await this.getIndex(index);
    if (!idx) {
      return;
    }

    return Array.from(idx.archives.keys());
  }

  public async getArchive(
    index: number,
    archive: number
  ): Promise<ArchiveData | undefined> {
    const idx = await this.getIndex(index);
    if (!idx) {
      return;
    }
    const am = idx.archives.get(archive);
    if (!am) {
      return;
    }
    if (!am.compressedData) {
      const d = await this.getArchiveRaw(index, archive);
      if (!d) {
        return;
      }
      am.compressedData = d;
    }
    return am;
  }

  private async getArchiveRaw(
    index: number,
    archive: number
  ): Promise<Uint8Array | undefined> {
    const ptrs = await this.getPointers(index);
    if (!ptrs) {
      return undefined;
    }

    ptrs.offset = archive * 6;
    if (ptrs.remaining <= 0) {
      return undefined;
    }

    const len = ptrs.u24();
    let sector = ptrs.u24();

    const data = await this.data;
    const r = new Reader(data!);

    const headerSize = archive > 0xffff ? 10 : 8;
    const SECTOR_SIZE = 520;

    const fastRead = len + headerSize <= SECTOR_SIZE;

    const obuf = fastRead ? undefined : new Uint8Array(len);
    let readBytes = 0;
    let part = 0;
    for (; readBytes < len; ) {
      r.offset = sector * SECTOR_SIZE;
      const readArchive = headerSize == 10 ? r.i32() : r.u16();
      const readPart = r.u16();
      const nextSector = r.u24();
      const readIndex = r.u8();

      if (readArchive != archive) {
        throw new Error(
          `corrupted; read archive ${readArchive} wanted ${archive}`
        );
      }
      if (readPart != part) {
        throw new Error(`corrupted; read part ${readPart} wanted ${part}`);
      }
      if (readIndex != index) {
        throw new Error(`corrupted; read index ${readIndex} wanted ${index}`);
      }

      const sa = r.array(Math.min(SECTOR_SIZE - headerSize, len - readBytes));
      if (fastRead) {
        return sa;
      }
      obuf!.set(sa, readBytes);
      readBytes += sa.length;

      sector = nextSector;
      part++;
    }

    return obuf;
  }

  public async getArchiveByName(
    index: number,
    name: string | number
  ): Promise<ArchiveData | undefined> {
    const namehash = hash(name);

    const idx = await this.getIndex(index);
    if (!idx) {
      return;
    }

    for (const ar of idx.archives.values()) {
      if (ar.namehash === namehash) {
        if (!ar.compressedData) {
          const d = await this.getArchiveRaw(index, ar.archive);
          if (!d) {
            return;
          }
          ar.compressedData = d;
        }
        return ar;
      }
    }
  }

  public async getVersion(index: number): Promise<CacheVersion> {
    return {
      era: "osrs",
      indexRevision: (await this.getIndex(index))?.revision ?? 0,
    };
  }
}
