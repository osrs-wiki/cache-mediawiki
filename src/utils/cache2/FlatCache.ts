import {
  ArchiveData,
  CacheProvider,
  CacheVersion,
  FileProvider,
  hash,
  IndexData,
} from "./Cache";

export class FlatIndexData implements IndexData {
  public revision!: number;
  public compression!: number;
  public crc!: number;
  public named!: boolean;

  /**@internal*/ archives: Map<number, ArchiveData> = new Map();
  private constructor(public id: number) {}

  static async of(indexID: number, data: Uint8Array): Promise<FlatIndexData> {
    const LF = 0x0a;
    const EQ = 0x3d;
    const fid = new FlatIndexData(indexID);
    const td = new TextDecoder();
    let ar: ArchiveData | undefined;
    for (let i = 0; i < data.length; i++) {
      let eol = data.indexOf(LF, i);
      if (eol == -1) {
        eol = data.length;
      }
      const line = data.subarray(i, eol);
      i = eol;

      const kvs = line.indexOf(EQ);
      if (kvs == -1) {
        continue;
      }
      const key = td.decode(line.subarray(0, kvs));
      const value = td.decode(line.subarray(kvs + 1));

      if (key === "id") {
        fid.archives.set(~~value, (ar = new ArchiveData(indexID, ~~value)));
        continue;
      }

      if (!ar) {
        switch (key) {
          case "named":
            fid.named = value == "true";
            break;
          case "revision":
          case "compression":
          case "crc":
            fid[key] = ~~value;
            break;
        }
      } else {
        switch (key) {
          case "contents": {
            const res = await fetch(
              "data:application/octet-binary;base64," + value
            );
            const data = await res.arrayBuffer();
            ar!.compressedData = new Uint8Array(data);
            break;
          }
          case "file": {
            const [id, hash] = value.split("=");
            ar!.addFile(~~id, ~~hash);
            break;
          }
          case "namehash":
          case "revision":
          case "crc":
          case "compression":
            ar[key] = ~~value;
            break;
        }
      }
    }
    return fid;
  }

  getArchive(archive: number): ArchiveData | undefined {
    return this.archives.get(archive);
  }

  getArchiveByName(name: string | number): ArchiveData | undefined {
    const namehash = hash(name);
    for (const ar of this.archives.values()) {
      if (ar.namehash == namehash) {
        return ar;
      }
    }
  }
}

export class FlatCacheProvider implements CacheProvider {
  private indexes: Map<number, Promise<FlatIndexData | undefined>> = new Map();

  constructor(private disk: FileProvider) {}

  public async getIndex(index: number): Promise<FlatIndexData | undefined> {
    let idxp = this.indexes.get(index);
    if (!idxp) {
      this.indexes.set(
        index,
        (idxp = this.disk
          .getFile(index + ".flatcache")
          .then((data) => data && FlatIndexData.of(index, data)))
      );
    }
    return idxp;
  }

  public async getArchive(
    index: number,
    archive: number
  ): Promise<ArchiveData | undefined> {
    const idx = await this.getIndex(index);
    return idx?.getArchive(archive);
  }

  public async getArchives(index: number): Promise<number[] | undefined> {
    const idx = await this.getIndex(index);
    if (!idx) {
      return;
    }

    return Array.from(idx.archives.keys());
  }

  public async getArchiveByName(
    index: number,
    name: string | number
  ): Promise<ArchiveData | undefined> {
    const idx = await this.getIndex(index);
    return idx?.getArchiveByName(name);
  }

  public async getVersion(index: number): Promise<CacheVersion> {
    return {
      era: "osrs",
      indexRevision: (await this.getIndex(index))?.revision ?? 0,
    };
  }
}
