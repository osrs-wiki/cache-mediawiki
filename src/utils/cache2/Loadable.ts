import { CacheProvider } from "./Cache";
import { Reader } from "./Reader";

type LoadableType<I extends Loadable, ARGS extends unknown[]> = {
  decode(reader: Reader, ...args: ARGS): I;
  loadData(cache: CacheProvider, ...args: ARGS): Promise<Reader | undefined>;
};

type OrNumber<T extends unknown[]> = T;

export abstract class Loadable {
  public static load<I extends Loadable, ARGS extends unknown[]>(
    this: LoadableType<I, ARGS>,
    cache: CacheProvider | Promise<CacheProvider>,
    ...args: OrNumber<ARGS>
  ): Promise<I | undefined>;
  public static load<I extends Loadable, ARGS extends unknown[]>(
    this: LoadableType<I, ARGS>,
    reader: Reader | ArrayBufferView | ArrayBuffer,
    ...args: OrNumber<ARGS>
  ): I;
  public static load<I extends Loadable, ARGS extends unknown[]>(
    this: LoadableType<I, ARGS>,
    reader:
      | Reader
      | ArrayBufferView
      | ArrayBuffer
      | CacheProvider
      | Promise<CacheProvider>,
    ...args: OrNumber<ARGS>
  ): Promise<I | undefined> | I {
    if (reader instanceof ArrayBuffer || ArrayBuffer.isView(reader)) {
      reader = new Reader(reader);
    }
    if (reader instanceof Reader) {
      return this.decode(reader, ...args);
    } else {
      return (async () => {
        const data = await this.loadData(await reader, ...args);
        if (data) {
          return this.decode(data, ...args);
        }
      })();
    }
  }
}

export abstract class PerArchiveLoadable extends Loadable {
  public static async loadData(
    this: { index: number },
    cache: CacheProvider,
    id: number
  ): Promise<Reader | undefined> {
    const archive = await cache.getArchive(this.index, id as number);
    const version = await cache.getVersion(this.index);
    const data = archive?.getFile(0)?.data;
    return data ? new Reader(data, version) : undefined;
  }

  public static async all<I extends PerArchiveLoadable, ID extends number>(
    this: {
      index: number;
      decode(reader: Reader, id: ID): I;
    },
    cache0: CacheProvider | Promise<CacheProvider>
  ): Promise<I[]> {
    const cache = await cache0;
    const ids = await cache.getArchives(this.index);
    if (!ids) {
      return [];
    }

    const archives = await Promise.all(
      ids.map((id) => cache.getArchive(this.index, id))
    );
    const version = await cache.getVersion(this.index);

    return archives
      .filter((v) => v)
      .map((v) => {
        try {
          return this.decode(
            new Reader(v!.getFile(0)!.data, version),
            v!.archive as ID
          );
        } catch (e) {
          if (typeof e === "object" && e && "message" in e) {
            const ea = e as any;
            ea.message = v!.archive + ": " + ea.message;
          }
          throw e;
        }
      });
  }
}

export abstract class NamedPerArchiveLoadable extends PerArchiveLoadable {
  public static async loadByName<
    I extends PerArchiveLoadable,
    ID extends number
  >(
    this: {
      index: number;
      decode(reader: Reader, id: ID): I;
    },
    cache0: CacheProvider | Promise<CacheProvider>,
    name: string | number
  ): Promise<I | undefined> {
    const cache = await cache0;
    const ar = await cache.getArchiveByName(this.index, name);
    const version = await cache.getVersion(this.index);
    const data = ar?.getFile(0)?.data;
    if (data) {
      return this.decode(new Reader(data, version), ar!.archive as ID);
    }
  }
}

export class PerFileLoadable extends Loadable {
  public static async loadData(
    this: { index: number; archive: number },
    cache: CacheProvider,
    id: number
  ): Promise<Reader | undefined> {
    const archive = await cache.getArchive(this.index, this.archive);
    const version = await cache.getVersion(this.index);
    const data = archive?.getFile(id)?.data;
    return data ? new Reader(data, version) : undefined;
  }

  public static async all<I extends PerFileLoadable, ID extends number>(
    this: {
      index: number;
      archive: number;
      decode(reader: Reader, id: ID): I;
    },
    cache0: CacheProvider | Promise<CacheProvider>
  ): Promise<I[]> {
    const cache = await cache0;
    const ad = await cache.getArchive(this.index, this.archive);
    if (!ad) {
      return [];
    }

    const version = await cache.getVersion(this.index);

    return [...ad.getFiles().values()]
      .filter((v) => v.data.length > 1 && v.data[0] != 0)
      .map((v) => {
        try {
          return this.decode(new Reader(v.data, version), v.id as ID);
        } catch (e) {
          if (typeof e === "object" && e && "message" in e) {
            const ea = e as any;
            ea.message = v.id + ": " + ea.message;
          }
          throw e;
        }
      });
  }
}
