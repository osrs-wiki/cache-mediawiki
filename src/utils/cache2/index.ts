export * from "./LazyPromise";
export * from "./types";
