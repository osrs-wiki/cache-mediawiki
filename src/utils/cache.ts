import { DiskCacheProvider, FlatCacheProvider } from "@abextm/cache2";
import { readFile } from "fs/promises";

export type CacheSource = "github" | "local";
export type CacheFileType = "flat" | "disk";

export const getCacheProviderLocal = async (
  version: string,
  type: CacheFileType
) => {
  if (type === "disk") {
    return getCacheProviderDiskLocal(version);
  } else if (type === "flat") {
    return getCacheProviderFlatLocal(version);
  }
};

export const getCacheProviderDiskLocal = async (version: string) => {
  return new DiskCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/diskcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderFlatLocal = async (version: string) => {
  return new FlatCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/flatcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderGithub = async (version = "master") => {
  return new FlatCacheProvider({
    async getFile(name) {
      const req = await fetch(
        `https://raw.githubusercontent.com/abextm/osrs-cache/${version}/${name}`
      );
      if (!req.ok) {
        return;
      }
      const ab = await req.arrayBuffer();
      return new Uint8Array(ab);
    },
  });
};
