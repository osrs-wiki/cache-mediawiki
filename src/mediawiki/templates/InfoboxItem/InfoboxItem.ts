import {
  InfoboxTemplate,
  MediaWikiDate,
  MediaWikiFile,
} from "@osrs-wiki/mediawiki-builder";
import type { InfoboxItem as InfoboxItemTemplate } from "@osrs-wiki/mediawiki-builder";

import { getInventoryActions } from "./InfoboxItem.utils";
import Context from "../../../context";
import { Item } from "../../../utils/cache2";

const InfoboxItem = (item: Item) => {
  return new InfoboxTemplate<InfoboxItemTemplate>("item", {
    name: item.name as string,
    image: new MediaWikiFile(`${item.name}.png`),
    release: Context.updateDate
      ? new MediaWikiDate(new Date(Context.updateDate))
      : undefined,
    update: Context.update,
    members: item.isMembers as boolean,
    quest: "No",
    exchange: item.isGrandExchangable,
    tradeable: item.isGrandExchangable,
    bankable: item.placeholderLinkedItem > 0,
    placeholder: item.placeholderLinkedItem > 1 ? true : undefined,
    equipable: item.wearpos1 > 0,
    stackable: item.isStackable,
    noteable: item.noteLinkedItem > 0,
    options: getInventoryActions(item),
    examine: item.examine,
    value: item.price,
    weight: (item.weight / 1000).toFixed(3),
    id: item.id.toString(),
  });
};

export default InfoboxItem;
