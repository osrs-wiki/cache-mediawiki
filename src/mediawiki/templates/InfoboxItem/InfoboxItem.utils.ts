import { MediaWikiFile, MediaWikiHTML } from "@osrs-wiki/mediawiki-builder";

import { Item } from "@/utils/cache2";

/**
 * Get the inventory actions for a given item.
 * @param item The item to get inventory actions for.
 * @returns An array of inventory actions, possibly with sub-operations included.
 */
export const getInventoryActions = (item: Item) => {
  return item.inventoryActions?.map((action, index) => {
    const subops = item.subops?.[index];
    return subops?.length > 0 ? `${action} (${subops.join(", ")})` : action;
  });
};

/**
 * Get the item infobox image for a given item, with support for stack variants.
 * If the item has no stack variants, it returns a single image.
 * If it has stack variants, it returns an array of images, inserting a break
 * after the first half if there are 5 or more variants.
 * @param item The item to get the infobox image for.
 * @returns The infobox image for the item.
 */
export const getItemInfoboxImage = (item: Item) => {
  const filteredVariants = item.stackVariantItems.filter((id) => id !== 0);
  if (filteredVariants.length === 0) {
    return new MediaWikiFile(`${item.name}.png`);
  } else {
    const variantFiles = filteredVariants.map(
      (_variantId, index) =>
        new MediaWikiFile(
          `${item.name} ${item.stackVariantQuantities[index]}.png`
        )
    );
    if (variantFiles.length >= 5) {
      const half = Math.floor(variantFiles.length / 2);
      return [
        ...variantFiles.slice(0, half),
        new MediaWikiHTML("br"),
        ...variantFiles.slice(half),
      ];
    } else {
      return variantFiles;
    }
  }
};
