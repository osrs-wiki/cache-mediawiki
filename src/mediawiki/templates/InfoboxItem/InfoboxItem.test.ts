import InfoboxItem from "./InfoboxItem";

import { Item, ItemID, Params, WearPos } from "@/utils/cache2";

const BASE_ITEM: Item = {
  name: "Test Item",
  isMembers: true,
  isGrandExchangable: true,
  id: 1234 as ItemID,
  inventoryModel: undefined,
  examine: "Examine text.",
  zoom2d: 0,
  xan2d: 0,
  yan2d: 0,
  offsetX2d: 0,
  offsetY2d: 0,
  isStackable: false,
  price: 0,
  wearpos1: undefined,
  wearpos2: undefined,
  wearpos3: undefined,
  weight: 0,
  maleModel: undefined,
  maleOffset: 0,
  maleModel1: undefined,
  femaleModel: undefined,
  femaleOffset: 0,
  femaleModel1: undefined,
  groundActions: [],
  inventoryActions: ["Test Action"],
  subops: [],
  recolorFrom: [],
  recolorTo: [],
  retextureFrom: [],
  retextureTo: [],
  shiftClickIndex: 0,
  maleModel2: undefined,
  femaleModel2: undefined,
  maleChatheadModel: undefined,
  femaleChatheadModel: undefined,
  maleChatheadModel2: undefined,
  femaleChatheadModel2: undefined,
  category: undefined,
  zan2d: 0,
  noteLinkedItem: undefined,
  noteTemplate: undefined,
  stackVariantItems: [],
  stackVariantQuantities: [],
  resizeX: 0,
  resizeY: 0,
  resizeZ: 0,
  ambient: 0,
  contrast: 0,
  team: 0,
  noted2: undefined,
  noted3: undefined,
  placeholderLinkedItem: undefined,
  placeholderTemplate: undefined,
  params: new Params(),
};

describe("IfnoboxItem", () => {
  test("InfoboxItem - base", () => {
    const item = InfoboxItem(BASE_ITEM);
    expect(item).toMatchSnapshot();
  });

  test("InfoboxItem - with wearpos1 >= 0", () => {
    const item = InfoboxItem({ ...BASE_ITEM, wearpos1: WearPos.Head });
    expect(item).toMatchSnapshot();
  });

  test("InfoboxItem - with stack variants", () => {
    const item = InfoboxItem({
      ...BASE_ITEM,
      stackVariantItems: [
        1 as ItemID,
        2 as ItemID,
        3 as ItemID,
        4 as ItemID,
        5 as ItemID,
      ],
      stackVariantQuantities: [10, 20, 30, 40, 50],
    });
    expect(item).toMatchSnapshot();
  });

  test("InfoboxItem - with stack variants that are all 0", () => {
    const item = InfoboxItem({
      ...BASE_ITEM,
      stackVariantItems: [0 as ItemID, 0 as ItemID, 0 as ItemID],
      stackVariantQuantities: [0, 0, 0],
    });
    expect(item).toMatchSnapshot();
  });
});
