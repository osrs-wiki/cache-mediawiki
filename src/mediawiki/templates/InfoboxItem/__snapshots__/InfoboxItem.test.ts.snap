// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IfnoboxItem InfoboxItem - base 1`] = `
InfoboxTemplate {
  "name": "Infobox Item",
  "params": Object {
    "bankable": false,
    "equipable": false,
    "examine": "Examine text.",
    "exchange": true,
    "id": "1234",
    "image": MediaWikiFile {
      "children": undefined,
      "fileName": "Test Item.png",
      "options": undefined,
    },
    "members": true,
    "name": "Test Item",
    "noteable": false,
    "options": Array [
      "Test Action",
    ],
    "placeholder": undefined,
    "quest": "No",
    "release": undefined,
    "stackable": false,
    "tradeable": true,
    "update": undefined,
    "value": 0,
    "weight": "0.000",
  },
}
`;

exports[`IfnoboxItem InfoboxItem - with wearpos1 >= 0 1`] = `
InfoboxTemplate {
  "name": "Infobox Item",
  "params": Object {
    "bankable": false,
    "equipable": true,
    "examine": "Examine text.",
    "exchange": true,
    "id": "1234",
    "image": MediaWikiFile {
      "children": undefined,
      "fileName": "Test Item.png",
      "options": undefined,
    },
    "members": true,
    "name": "Test Item",
    "noteable": false,
    "options": Array [
      "Test Action",
    ],
    "placeholder": undefined,
    "quest": "No",
    "release": undefined,
    "stackable": false,
    "tradeable": true,
    "update": undefined,
    "value": 0,
    "weight": "0.000",
  },
}
`;
